framework/src/modules/dpos/dpos_module.ts

```diff
@@ -12,90 +12,223 @@
+		if (errors.length) {
+			throw new LiskValidationError([...errors]);
+		}
+
+		this._activeDelegates = this.config.activeDelegates as number;
+		this._standbyDelegates = this.config.standbyDelegates as number;
+		this._delegateListRoundOffset = this.config.delegateListRoundOffset as number;
+		this._delegatesPerRound = this._activeDelegates + this._standbyDelegates;
+		this._blockTime = config.blockTime;
+		this._delegateActiveRoundLimit = 3;
+
+		this.rounds = new Rounds({ blocksPerRound: this._delegatesPerRound });
+	}
+
+	// eslint-disable-next-line consistent-return
+	public async afterBlockApply(input: AfterBlockApplyInput): Promise<void> {
+		const finalizedHeight = input.consensus.getFinalizedHeight();
+		const lastBootstrapHeight = input.consensus.getLastBootstrapHeight();
+		const isLastBlockOfRound = this._isLastBlockOfTheRound(input.block.header);
+		const { height } = input.block.header;
+		const isBootstrapPeriod = height <= lastBootstrapHeight;
+
+		if (finalizedHeight !== this._finalizedHeight) {
+			this._finalizedHeight = finalizedHeight;
+
+			const finalizedBlockRound = this.rounds.calcRound(finalizedHeight);
+			const disposableDelegateListUntilRound =
+				finalizedBlockRound - this._delegateListRoundOffset - this._delegateActiveRoundLimit;
+			await deleteVoteWeightsUntilRound(disposableDelegateListUntilRound, input.stateStore);
+		}
+
+		if (!isBootstrapPeriod) {
+			// Calculate account.dpos.delegate.consecutiveMissedBlocks and account.dpos.delegate.isBanned
+			await this._updateProductivity(input);
+		}
+
+		if (!isLastBlockOfRound) {
+			return Promise.resolve();
+		}
+
+		await this._createVoteWeightSnapshot(input);
+
+		if (!isBootstrapPeriod || (isBootstrapPeriod && height === lastBootstrapHeight)) {
+			await this._updateValidators(input);
+		}
+	}
+
+	// eslint-disable-next-line @typescript-eslint/explicit-member-accessibility,class-methods-use-this,@typescript-eslint/require-await
+	public async afterGenesisBlockApply<T = Account<DPOSAccountProps>>(
+		input: AfterGenesisBlockApplyInput<T>,
+	): Promise<void> {
+		const { accounts, initDelegates } = input.genesisBlock.header.asset;
+		const delegateAddresses: Buffer[] = [];
+		const delegateUsernames: { address: Buffer; username: string }[] = [];
+
+		for (const account of (accounts as unknown) as Account<DPOSAccountProps>[]) {
+			if (account.dpos.delegate.username !== '') {
+				delegateUsernames.push({
+					address: account.address,
+					username: account.dpos.delegate.username,
+				});
+				delegateAddresses.push(account.address);
+			}
+		}
+
+		if (!bufferArrayUniqueItems([...initDelegates])) {
+			throw new Error('Genesis block init delegates list contains duplicate addresses');
+		}
+
+		if (!bufferArrayOrderByLex([...initDelegates])) {
+			throw new Error('Genesis block init delegates list is not ordered lexicographically');
+		}
+
+		if (initDelegates.length > this._delegatesPerRound) {
```

Comment:
I think this should be updated to `!==`

Reply:
Init delegates list can be less than round length. It can even contain only 1 delegate, but it should not contain more than what is in the round length.